go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,482 @@
+
+### WorkflowInboundInterceptor.HandleSignal
+
+Current:
+
+```go
+ProcessSignal(ctx Context, signalName string, arg interface{}) error
+```
+
+Proposal:
+
+```go
+type SignalInput struct {
+  SignalName string
+  Arg interface{}
+}
+
+HandleSignal(ctx Context, input *SignalInput) error
+```
+
+Also, this will be called to handle the signal with the last invocation in the chain passing the signal to the workflow.
+
+Why:
+
+* `HandleSignal` matches Java nomenclature and new behavior better
+* Currently the sending of the signal to the handler is done after the interceptor not allowing the interceptor to
+  actually intercept and alter/prevent the send. We will change to have the last interceptor do the actual send.
+* We will need an integration test to confirm that this is invoked on start-with-signal workflow that has a sync receive
+  as the first statement
+
+### WorfkflowInboundInterceptor.HandleQuery
+
+Current:
+
+```go
+HandleQuery(ctx Context, queryType string, args *commonpb.Payloads,
+  handler func(*commonpb.Payloads) (*commonpb.Payloads, error)) (*commonpb.Payloads, error)
+```
+
+Proposal:
+
+```go
+type QueryInput struct {
+  QueryType string
+  Args []interface{}
+}
+
+HandleQuery(ctx Context, input *QueryInput) (interface{}, error)
+```
+
+Why:
+
+* All other calls convert the payload outside of the interceptor call, this can too
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
+
+### WorkflowOutboundInterceptor
+
+We can leave everything the way it is because, while it doesn't make the interceptor parameters future-proof with struct
+wrappers, it mimics the actual calls that aren't going to have backwards-compatible alterations.
+
+### WorkflowOutboundInterceptor.GetInfo
+
+Current:
+
+```go
+GetWorkflowInfo(ctx Context) *WorkflowInfo
+```
+
+Proposal:
+
+```go
+GetInfo(ctx Context) *WorkflowInfo
+```
+
+Why: All other calls match `workflow` package, why not this?
+
+Note: We need to make sure we document that this returns *workflow.Info since internal names can be confusing.
+
+### WorkerInterceptor.InterceptActivity
+
+Proposal:
+
+```go
+InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+```
+
+Why: Mimics `InterceptWorkflow` but for activities.
+
+### ActivityInboundInterceptor
+
+Proposal:
+
+```go
+type ActivityInput struct {
+  Args []interface{}
+}
+
+type ActivityInboundInterceptor interface {
+  Init(outbound ActivityOutboundInterceptor) error
+
+  ExecuteActivity(ctx context.Context, input *ActivityInput) (interface{}, error)
+
+  mustEmbedActivityInboundInterceptorBase()
+}
+```
+
+Also a `ActivityInboundInterceptorBase` implementation with expected delegation.
+
+Why: Mimics `WorkflowInboundInterceptor` but for activities.
+
+### ActivityOutboundInterceptor
+
+Proposal:
+
+```go
+type ActivityOutboundInterceptor interface {
+  GetInfo(ctx context.Context) ActivityInfo
```

Comment:
Are we concerned about a the costs of copying the result on every invocation in the chain? Might we want to return a pointer here instead (even though the last call can dereference)?

Reply:
I don't believe we are concerned with the copy